### DISCUSSION TITLE
File Browser upgrade

### DISCUSSION THREAD

**Discussion by grch87** [biologist79/ESPuino#49] [General]
Hi,
as my ESP Tonuino is finished, I got the request from friends, if I can build them as well one. However, they don't feel confident enough to deal with an FTP client to upload the files.
Therefore I decided to extend the current file browser with the possibility to upload files and add some basic functionalities:
- Create Folder
- Delete File/Folder (nested)
- Rename File/Folder
- Refresh
- PlayAudio

I used the current solution with jstree, but had to change the way how the data of the filesystem is synchronized between the ESP and the client. 
I tried to use a more dynamic approach. Instead of one big json file, containing all files and folders of the filesystem, the client is requesting the content of a folder from the ESP, when the folder is entered or something got changed in the folder (e.g. upload, deletion...).

I perfomed the commit under:
https://github.com/grch87/Tonuino-ESP32-I2S/commit/5ba8fb4ecef713ee53f9e299c33434ec555af8df

And here a picture how it looks like:

<img width="361" alt="FileBrowser" src="https://user-images.githubusercontent.com/70661090/103821642-ebf10d00-506e-11eb-9177-340ed1433d99.png">

Maybe this is as well a nice feature for the master branch.

I just faced one issue. I store the html file on the sdcard which is working fine. When I use the processHtml script to integrate the html code into the source code the html page is not properly shown on the client side. Any idea? Maybe the page got too big?

**Comment by biologist79** [comment]
Wow, cool! Will have a look at it.
Regarding the problem... rust a rough guess: Have to been commenting out stuff using //? Problem is that the whole html-file is finally a one-liner. So commenting out can lead to commenting out all subsequent stuff :-) And yes, I ran into this problem, hehe.

**Comment by biologist79** [comment]
I got it running using send_p(), but it's to fully functional. So yes, maybe it's way to big now.
However, when serving via SD, you also need the template-processor:
`request->send(FSystem, "/index.html", String(), false, templateProcessor);`

Tested the upload-function. Is around 140 kB/s. Way slower than FTP, but it's working anyway :-)

**Comment by biologist79** [comment]
Cherry-picked your commit into a new feature-branch: https://github.com/biologist79/Tonuino-ESP32-I2S/tree/feature_filebrowser

**Reply by grch87** [reply]
Moin,
I fixed the problem, that the website couldn't be loaded from the ROM:
https://github.com/grch87/Tonuino-ESP32-I2S/commit/2ac749863284834c64c06d1cd27677647f77b8ab

Reason was that some semicolons were missing. This causes of course problems when everything is in one line :D

**Comment by grch87** [comment]
Hi,
I did an update in the upload part. I moved the writing of the data onto the SD card to another task. So aquisition of the data via the network and SD storage is now done in parallel, which increases the speed by 50% (280-360kB/s jitters quite a lot). I know it is increasing the complexity, but for the speed increase it is worth I think.
My commit is here:
https://github.com/grch87/Tonuino-ESP32-I2S/commit/e429970d39356c16d26de7b7daaa6dc094bc9789

There was as well a bug. The sleep timer is not reseted during the upload, so the tonuino can go into sleep. This is as well now fixed with this commit.

Cheers

**Comment by biologist79** [comment]
Well done, integrated the commit yesterday. But still had an issue when uploading large files as Tonuino went to sleep after the period configured.
